File: hh-examples/04-2-collection-jpa/src/main/java/com/oreilly/hh/data/Track.java
package com.oreilly.hh.data;

import java.time.LocalDate;
import java.time.LocalTime;
import java.util.HashSet;
import java.util.Set;

import jakarta.persistence.*;

@NamedQuery(name = "com.oreilly.hh.tracksNoLongerThan",
            query = """
                    SELECT track
                    FROM Track AS track
                    WHERE track.playTime <= :length
                    """)

@Entity

@lombok.Data
@lombok.EqualsAndHashCode(of="id")
@lombok.ToString(exclude={"playTime", "added", "volume", "comments"})
public class Track  implements java.io.Serializable {

    @Id
    @Column(name = "TRACK_ID")
    @GeneratedValue(strategy = GenerationType.AUTO)
    private Integer id;

    private String title;

    private String filePath;

    private LocalTime playTime = LocalTime.of(0, 0, 0);

    private LocalDate added = LocalDate.now();

    private short volume = 0;

    @ElementCollection
    @JoinTable(name = "TRACK_COMMENTS",
               joinColumns = @JoinColumn(name = "TRACK_ID"))
    @Column(name = "COMMENT")
    private Set<String> comments = new HashSet<String>();

    @ManyToMany
    @JoinTable(name = "TRACK_ARTISTS",
               joinColumns = {@JoinColumn(name = "TRACK_ID")},
               inverseJoinColumns = {@JoinColumn(name = "ARTIST_ID")})
    private Set<Artist> artists = new HashSet<Artist>();

    public Track() {
    }

    public Track(String title, String filePath, LocalTime playTime) {
        this.title = title;
        this.filePath = filePath;
        this.playTime = playTime;
    }

    public Track(String title, String filePath, LocalTime playTime, Set<Artist> artists, short volume, Set<String> comments) {
        this.title = title;
        this.filePath = filePath;
        this.playTime = playTime;
        this.artists = artists;
        this.volume = volume;
        this.comments = comments;
    }

}

File: hh-examples/02-4-mapping-jpa-kotlin/src/main/kotlin/com/oreilly/hh/data/Track.kt
package com.oreilly.hh.data

import java.time.LocalDate
import java.time.LocalTime

import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.GeneratedValue
import jakarta.persistence.Id
import jakarta.persistence.NamedQuery

@NamedQuery(
        name = "com.oreilly.hh.tracksNoLongerThan",
        query = "SELECT track FROM Track as track WHERE track.playTime <= :length")

@Entity
class Track(
    var title: String = "",

    var filePath: String = "",

    var playTime: LocalTime = LocalTime.of(0, 0, 0),

    var added: LocalDate = LocalDate.now(),

    var volume: Short = 0,

    @Id
    @Column(name = "TRACK_ID")
    @GeneratedValue
    var id: Int = 0
) {
    override fun toString(): String = "Track(id=$id, title='$title')"
    override fun equals(other: Any?): Boolean = other is Track && other.id == this.id
    override fun hashCode(): Int = id
}

